### ISSUE TITLE
[question] Alternative for cmake_paths generator

### ISSUE THREAD

**Issue by grabusr** [conan-io/conan#16108]
### What is your question?

Hi

We would like to use CMake Configs provided by library. For instance google test library has `install` target which put library files, header files to `CMAKE_INSTALL_PREFIX` directory, but also `GTestConfig.cmake`, `GTestConfigVersion.cmake`, which we would like to use. I created recipe:
```python
from conan import ConanFile, tools
from conan.tools.cmake import CMakeToolchain, CMake
from conan.tools.scm import Git

class GoogleTestConnan(ConanFile):
    name = "qts-gtest"
    version = "1.14.0"
    settings = "os", "arch", "compiler", "build_type"
    url = "https://github.com/google/googletest.git"

    def source(self):
        git = Git(self)
        git.clone(self.url, target=".")
        git.checkout("v" + self.version)

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package_info(self):
        self.cpp_info.set_property("cmake_find_mode", "none")

    def package(self):
        cmake = CMake(self)
        cmake.install()

    def generate(self):
        tc = CMakeToolchain(self)
        tc.variables["BUILD_GMOCK"] = True
        tc.generate()

```
When I used `cmake_paths` generator the we could just put given code in root `CMakeLists.txt`:
```cmake
if (EXISTS ${CMAKE_BINARY_DIR}/conan_paths.cmake)
    include(${CMAKE_BINARY_DIR}/conan_paths.cmake)
endif ()
```
But now `cmake_paths` generator is deprecated. How I can now generate the file with this:
```
set(CMAKE_PREFIX_PATH "<home>/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" <other libs...>
```
Now when I use `CMakeToolchain` conan generates `conan_toolchain.cmake` file, but there is only one `CMAKE_PREFIX_PATH` modification:
```
list(PREPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_LIST_DIR} )
```
which does not point to any package. 
There is line:
```
list(PREPEND CMAKE_MODULE_PATH "<home>/.conan/data/gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" <other-packages>
```
Maybe it is a bug, because having `CMAKE_PREFIX_PATH` instead of `CMAKE_MODULE_PATH` solves the solution.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @grabusr 

Thanks for your question.

The functionality of ``cmake_paths`` has been implemented in the new ``CMakeToolchain``.

There is a full working example of this use case in the docs: https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/use_package_config_cmake.html, please have a look and let us know.

**Comment by grabusr** [comment]
Hi @memsharded

Yeah, I went though that. My `conanfile.py` is:
```
from conans.model.conan_file import ConanFile

class MyConan(ConanFile):
    settings = "os", "compiler", "build_type", "arch"
    requires = "qts-gtest/1.14.0"
    generators = "CMakeToolchain"
```
I run command from my build directory
```bash
conan install .. --output-folder=.
```
but as I said `CMAKE_PREFIX_PATH` does not contain path to `GTest` package.

When I "debug" it like this:
```cmake
message("CMAKE_PREFIX_PATH: ${CMAKE_PREFIX_PATH}")
find_package(GTest REQUIRED)
```
The `CMAKE_PREFIX_PATH` contains only path to my build directory.

**Comment by memsharded** [comment]
How do you call cmake? You need to pass the toolchain (via -DCMAKE_TOOLCHAIN_FILE or using the generated conan CMake presets).

**Comment by memsharded** [comment]
> from conans.model.conan_file import ConanFile

This is broken. Please use only imports ``from conan``, not ``from conans``.

**Comment by grabusr** [comment]
>This is broken. Please use only imports from conan, not from conans.

I replaced `from conans` by `from conan`, but nothing has changed.

>How do you call cmake? You need to pass the toolchain (via -DCMAKE_TOOLCHAIN_FILE or using the generated conan CMake presets).

I insert to root CMakeLists.txt
```cmake
if (EXISTS ${CMAKE_BINARY_DIR}/conan_toolchain.cmake)
    include(${CMAKE_BINARY_DIR}/conan_toolchain.cmake)
endif ()
```
I tried also solution with `-DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake` but this not fix issue.

Having build directory only in `CMAKE_PREFIX_PATH` seems to be the reason because generated `conan_toolchain.cmake` is:

```cmake


# Conan automatically generated toolchain file
# DO NOT EDIT MANUALLY, it will be overwritten

# Avoid including toolchain file several times (bad if appending to variables like
#   CMAKE_CXX_FLAGS. See https://github.com/android/ndk/issues/323
include_guard()

message(STATUS "Using Conan toolchain: ${CMAKE_CURRENT_LIST_FILE}")

if(${CMAKE_VERSION} VERSION_LESS "3.15")
    message(FATAL_ERROR "The 'CMakeToolchain' generator only works with CMake >= 3.15")
endif()










string(APPEND CONAN_CXX_FLAGS " -m64")
string(APPEND CONAN_C_FLAGS " -m64")
string(APPEND CONAN_SHARED_LINKER_FLAGS " -m64")
string(APPEND CONAN_EXE_LINKER_FLAGS " -m64")



# Extra c, cxx, linkflags and defines


if(DEFINED CONAN_CXX_FLAGS)
  string(APPEND CMAKE_CXX_FLAGS_INIT " ${CONAN_CXX_FLAGS}")
endif()
if(DEFINED CONAN_C_FLAGS)
  string(APPEND CMAKE_C_FLAGS_INIT " ${CONAN_C_FLAGS}")
endif()
if(DEFINED CONAN_SHARED_LINKER_FLAGS)
  string(APPEND CMAKE_SHARED_LINKER_FLAGS_INIT " ${CONAN_SHARED_LINKER_FLAGS}")
endif()
if(DEFINED CONAN_EXE_LINKER_FLAGS)
  string(APPEND CMAKE_EXE_LINKER_FLAGS_INIT " ${CONAN_EXE_LINKER_FLAGS}")
endif()

get_property( _CMAKE_IN_TRY_COMPILE GLOBAL PROPERTY IN_TRY_COMPILE )
if(_CMAKE_IN_TRY_COMPILE)
    message(STATUS "Running toolchain IN_TRY_COMPILE")
    return()
endif()

set(CMAKE_FIND_PACKAGE_PREFER_CONFIG ON)

# Definition of CMAKE_MODULE_PATH
# The root (which is the default builddirs) path of dependencies in the host context
list(PREPEND CMAKE_MODULE_PATH "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" "/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" "/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" "/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/")
# the generators folder (where conan generates files, like this toolchain)
list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})

# Definition of CMAKE_PREFIX_PATH, CMAKE_XXXXX_PATH
# The Conan local "generators" folder, where this toolchain is saved.
list(PREPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_LIST_DIR} )
list(PREPEND CMAKE_PROGRAM_PATH "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/bin" "/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/bin" "/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/bin" "/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/bin")
list(PREPEND CMAKE_LIBRARY_PATH "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/lib" "/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/lib" "/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/lib" "/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/lib")
list(PREPEND CMAKE_INCLUDE_PATH "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/include" "/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/include" "/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/include" "/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/include")



if (DEFINED ENV{PKG_CONFIG_PATH})
set(ENV{PKG_CONFIG_PATH} "/home/grabowski/projects/DeckbuilderGame/cmake-build-debug-gcc:$ENV{PKG_CONFIG_PATH}")
else()
set(ENV{PKG_CONFIG_PATH} "/home/grabowski/projects/DeckbuilderGame/cmake-build-debug-gcc:")
endif()




# Variables
# Variables  per configuration


# Preprocessor definitions
# Preprocessor definitions per configuration

```

Moreover there is a lot of unnecessary things in `conan_toolchain.cmake` file comparing to `cmake_paths` generator generating pretty simple `conan_paths.cmake` file:
```cmake
set(CONAN_QTS-GTEST_ROOT "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9")
set(CONAN_QTS-NLOHMANN_JSON_ROOT "/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9")
set(CONAN_QTS-CATCH2_ROOT "/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9")
set(CONAN_QTS-GODOT_CPP_ROOT "/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9")
set(CMAKE_MODULE_PATH "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/"
			"/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/"
			"/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/"
			"/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" ${CONAN_QTS-GTEST_ROOT} ${CONAN_QTS-NLOHMANN_JSON_ROOT} ${CONAN_QTS-CATCH2_ROOT} ${CONAN_QTS-GODOT_CPP_ROOT} ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_LIST_DIR})
set(CMAKE_PREFIX_PATH "/home/grabowski/.conan/data/qts-gtest/1.14.0/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/"
			"/home/grabowski/.conan/data/qts-nlohmann_json/3.11.3/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/"
			"/home/grabowski/.conan/data/qts-catch2/3.5.4/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/"
			"/home/grabowski/.conan/data/qts-godot_cpp/4.2.1/_/_/package/062863c92a5a0a247840166e9f84ebe8d10786b9/" ${CONAN_QTS-GTEST_ROOT} ${CONAN_QTS-NLOHMANN_JSON_ROOT} ${CONAN_QTS-CATCH2_ROOT} ${CONAN_QTS-GODOT_CPP_ROOT} ${CMAKE_PREFIX_PATH} ${CMAKE_CURRENT_LIST_DIR})
```

**Comment by memsharded** [comment]
> include(${CMAKE_BINARY_DIR}/conan_toolchain.cmake)

Toolchains cannot be ``include()`` in CMakeLists.txt. They need to:
- be passed in command line as ``-DCMAKE_TOOLCHAIN_FILE=...``
- Added in CMake presets as ``CMAKE_TOOLCHAIN_FILE`` variable or as ``toolchainFile``


> Moreover there is a lot of unnecessary things in conan_toolchain.cmake file comparing to cmake_paths generator generating pretty simple conan_paths.cmake file:

You can control what the ``CMakeToolchain`` generates, it is fully customizable, you can leave the paths only https://docs.conan.io/2/reference/tools/cmake/cmaketoolchain.html#extending-and-advanced-customization

Also, this is important: those things are not unnecessary in most cases. They are very important to have a build that is consistent with what the Conan profile files have specified, and to be consistent with the dependencies builds, specially if they are built with other build systems, with happens often for open source libraries. For example if you pass ``-c tools.build:cxxflags=``, this is passed via this toolchain file, or if you specify ``-s compiler.cppstd=20``, it is also passed via this toolchain file. If you are not leveraging the toolchain file, all the customization and control of binaries is lost, can be incoherent (the Conan package_ids show one ``compiler.cppstd``, but the actual binary was built with a different ``compiler.cppstd`` that was hardcoded somewhere, and that might be different for every package without control).


Most likely the core issue is that you are using an older version of Conan. It seems that Conan 2.2 is working automatically fine by default, while older Conan 1.X versions might need to explicitly define the ``self.cpp_info.builddirs`` in the generating recipe.

It would be necessary a fully reproducible example, something like what we have in the examples2 repo explained in https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/use_package_config_cmake.html. If the above hint doesn't solve the issue, please try to put together a full repro case and we will try to help. Thanks!

**Comment by grabusr** [comment]
I investigated the generators python code. `cmake_paths` generator works because it just add path to root of dependency to `CMAKE_PREFIX_PATH`: https://github.com/conan-io/conan/blob/b7f925bbe66638fd406c0674132b04783a603cc8/conans/client/generators/cmake_paths.py#L35C45-L35C63

What is missing in our recipe to make it work with `CMakeToolchain` generator is adding one line to `def package_info(self)`:
```python
self.cpp_info.builddirs = ["."]
```
In my opinion it should be default behavior like in `cmake_paths` generator.

Please notice that you have a little mistake in setup `CMAKE_PREFIX_PATH` in [this example](https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/use_package_config_cmake.html), because you set:
```python
self.cpp_info.builddirs = ["pkg/cmake"]
```
which is path where is ConfigXXX.cmake, but following [CMake documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_PREFIX_PATH.html) `CMAKE_PREFIX_PATH` should point to directory with Unix like structure containing folders like bin, lib etc.

If you want to provide the direct paths to directories containing to FindXXX.cmake or ConfigXXX.cmake then you should use [CMAKE_MODULE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_MODULE_PATH.html) for that.

Summarizing discussion I think we will stay with `cmake_paths` generator. It is much simpler and easier to use. Just collect sources, run CMake configuration, run CMake build, run CMake install, no more additional settings and scripts are needed and there is no extra stuff like setting compiler flags in generated file.

Thank you for your help!

**Comment by memsharded** [comment]
> self.cpp_info.builddirs = ["."]
> In my opinion it should be default behavior like in cmake_paths generator.

There were problems in the past with packages injecting things unexpectedly, so this was made opt-in.

> Summarizing discussion I think we will stay with cmake_paths generator. It is much simpler and easier to use. Just collect sources, run CMake configuration, run CMake build, run CMake install, no more additional settings and scripts are needed and there is no extra stuff like setting compiler flags in generated file.

Let me clarify that ``cmake_paths`` is legacy, it is no longer supported. It was superseded by ``CMakeToolchain`` years ago, and it has already been fully removed from Conan 2.0, that was released more than 1 year ago after 2.5 years in beta, and it is now the recommended version for production. ConanCenter will soon (following months) stop receiving updates for Conan 1, that means that if you are using recipes from ConanCenter, your dependencies will be frozen, they won't break, but they won't receive any new fixes, new versions or new binaries at all. It is strongly recommended to upgrade to Conan 2 as soon as possible.


In any case, happy to know that the ``self.cpp_info.builddirs = ["."]`` worked, I am closing this ticket as solved, thanks for the feedback!
